### PR TITLE
Efficient forward & backward

### PR DESCRIPTION
**Pull request by dbaranchuk**
* Fault-tolerant sequential forward and backward;
* Efficient forward and backward calls that process an input batch in parallel by splitting it on small sub-batches;
* Update server deployment scripts.
---

Some measurements:

**Config:** Bloom 6b3 | input_ids.shape = (32, 128) | A100 | 3 servers, 10 block each

**Naive (current)**
84.2 sec / it | (1/3 GPUs is utilized at time)

**Sequential**
37.4 sec / it | (1/3 GPUs is utilized at time)

**Sequential + Parallel**
19.4 sec / it | (2/3 or 3/3 GPUs are utilized at time)

**Reference Gradient Checkpointing (Single GPU A100)**
8.8 sec / it | 100% gpu utils

**Reference (Single GPU A100)**
5.7 sec / it | 100% gpu utils

### CODE REVIEW

File: src/client/async_forward_backward.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        forward_sequences: List[Sequence[RemoteSpanInfo]] = ctx.sequences_for_batches
+        ctx.sequence_manager.rpc_info # lazy init
+
+        batch_size = max(MAX_TOKENS_IN_BATCH // grad_outputs.shape[1], 1)
+        grad_output_batches: Sequence[torch.Tensor] = grad_outputs.split(batch_size)
+        assert len(intermediate_input_batches) == len(grad_output_batches) == len(forward_sequences)
+
+        grad_input_batches = RemoteExpertWorker.run_coroutine(
+            _gather_backward(grad_output_batches, intermediate_input_batches, forward_sequences, ctx.sequence_manager)
+            # async_backward((grad_output_batches[0], ), intermediate_input_batches[0], forward_sequences[0], ctx.sequence_manager)
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:211]: [commented code, consider removing]

File: src/client/remote_model.py [hunk 2]
```diff
@@ -66,13 +67,39 @@ def set_requires_grad(self, value):
...
-            )
-        return super().forward(*args, use_cache=False, **kwargs)
+    def forward(
+        self, 
+        input_ids: Optional[torch.LongTensor] = None, 
+        inputs_embeds: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        **kwargs
+    ):
+        if input_ids is not None and inputs_embeds is not None:
```
Comment (justheuristic) [inline src/client/remote_model.py:77]: Consider adding a warning:

if kwargs:
  logger.warning("The following keyword arguments to DistributedBloom will will be ignored: {kwargs}")

File: src/client/remote_model.py [hunk 3]
```diff
@@ -66,13 +67,39 @@ def set_requires_grad(self, value):
...
-            raise ValueError(
-                "Distributed forward does not support use_cache; for efficient cache-aware generation, "
-                "please use model.transformer.inference_session() or model.generate(...)"
-            )
-        return super().forward(*args, use_cache=False, **kwargs)
+    def forward(
+        self, 
+        input_ids: Optional[torch.LongTensor] = None, 
+        inputs_embeds: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
```
Comment (justheuristic) [inline src/client/remote_model.py:74]: assert attention_mask is None

File: src/client/async_forward_backward.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                stub = TransformerConnectionHandler.get_stub(sequence_manager.p2p, span.peer_id)
+                
+                grad_outputs = await run_backward(
+                    span_uids, stub, sequence_manager.rpc_info, inputs, grad_outputs
+                )
+                break
+            except Exception as e:
+                logging.warning(f"Caught {e} when running backward for chain {span.start}-{span.end}", exc_info=True)
+                _, backup_intermediate_inputs, backup_forward_sequences = await async_forward(
+                    inputs, sequence_manager[span.start: span.end] # TODO: new sequence manager requires new rpc_info init and hence freezes
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:147]: TODO consider sending specific indices instead of slicing SequenceManager

File: src/client/async_forward_backward.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                )
+
+                forward_sequences = forward_sequences + backup_forward_sequences
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:150]: consider forward_sequences.extend

File: src/client/async_forward_backward.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+                # Save intermediate inputs and subsequences if the forward is already done for them
+                intermediate_inputs.append(inputs)
+                done_sequences.append(span)
+
+                inputs = outputs
+                break
+            except Exception as e:
+                logging.warning(f"Caught {e} when running forward for chain {span.start}-{span.end}", exc_info=True)
+                backup_sequences = sequence_manager[span.start: span.end].make_sequence()
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:113]: minor: consider make_sequence instead of lice

conceptual: consider making sequence from span_start to GLOBAL END (len(sequence_manager))
and replace sequence manager with this

File: src/client/async_forward_backward.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+                assert backup_sequences[0].start == span.start
+                assert backup_sequences[-1].end == span.end
+                sequences = backup_sequences + sequences[1:]
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:116]: .extend

File: src/client/async_forward_backward.py (hunk 1 above)
Comment (justheuristic) [inline src/client/async_forward_backward.py:211]: rm commented line

**Review by justheuristic — APPROVED** [review]

**Comment by justheuristic** [comment]
Please add a test case for test_remote_sequential when it actually splits tokens w.r.t. batches OR create an issue and tag me :)

File: src/client/async_forward_backward.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+MAX_TOKENS_IN_BATCH=1024
+
+
+async def run_forward(
+    uid: ModuleUID, 
+    stub: StubBase,
+    rpc_info: RPCInfo,
+    *inputs: torch.Tensor,
+    **kwargs
+) -> Tuple[torch.Tensor, ...]:
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:39]: Nit: consider adding a single-line docstring OR rename run_forward into something specific (so that users can distinguish it from async_forward, expert_forward)

File: src/client/async_forward_backward.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    forward_inputs = (inputs, kwargs)
+
+    if not nested_compare(forward_inputs, rpc_info["forward_schema"]):
+        raise TypeError(f"Inputs do not match expert input schema. Did you pass the right number of parameters?")
+ 
+    forward_inputs = nested_flatten(forward_inputs)
+    inputs = tuple(tensor.cpu().detach() for tensor in forward_inputs)
+
+    serialized_tensors = (
+        serialize_torch_tensor(tensor, proto.compression)
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:55]: [note to self: create a todo to figure out whether we should use run_in_executor]

**Comment by justheuristic** [comment]
[minor, reminder: i'd guess that the test_full_model will fix itself after merging #35 ]

File: src/client/async_forward_backward.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
+import logging
```
Comment (justheuristic) [inline src/client/async_forward_backward.py:1]: Nit: consider renaming into sequential_forward_backward or chain_autograd or similar
